verification/src/download_cache.rs

```diff
@@ -0,0 +1,206 @@
+use anyhow::Result;
+use async_trait::async_trait;
+use semver::Version;
+use std::{collections::HashMap, path::PathBuf, sync::Arc};
+
+#[async_trait]
+pub trait Fetcher {
+    async fn fetch(&self, ver: &Version) -> Result<PathBuf>;
+}
+
+pub struct DownloadCache<D> {
+    cache: parking_lot::Mutex<HashMap<Version, Arc<tokio::sync::RwLock<Option<PathBuf>>>>>,
+    downloader: D,
+}
+
+impl<D> DownloadCache<D> {
+    pub fn new(downloader: D) -> Self {
+        DownloadCache {
+            cache: Default::default(),
+            downloader,
+        }
+    }
+
+    async fn try_get(&self, ver: &Version) -> Option<PathBuf> {
+        let entry = {
+            let cache = self.cache.lock();
+            cache.get(ver).cloned()
+        };
+        match entry {
+            Some(lock) => {
+                let solc = lock.read().await;
+                solc.as_ref().cloned()
```

Comment:
since DownloadCache doesnt know anything about solc, I think it should not be `solc` variable